api/src/exam-environment/utils/exam.test.ts

```diff
@@ -0,0 +1,307 @@
+import { type Static } from '@fastify/type-provider-typebox';
+import { EnvExam } from '@prisma/client';
+import { exam, examAttempt, generatedExam } from '../../../__mocks__/env-exam';
+import * as schemas from '../schemas';
+import {
+  checkAttemptAgainstGeneratedExam,
+  constructUserExam,
+  generateExam,
+  userAttemptToDatabaseAttemptQuestionSets,
+  validateAttempt
+} from './exam';
+
+function timedoutGenerateExam(ex: EnvExam) {
+  const TIMEOUT_IN_MS = 1_000;
+  const START_TIME = Date.now();
+  // eslint-disable-next-line
+  while (true) {
+    if (Date.now() - START_TIME > TIMEOUT_IN_MS) {
+      throw `Unable to generate 1 valid exam in ${TIMEOUT_IN_MS}ms`;
+    }
+    try {
+      const exam = generateExam(ex);
+      return exam;
+    } catch (_e) {
+      //
+    }
+  }
+}
```

Comment:
I get that generateExam is random, but the tests shouldn't be. If you need random number generation in a test, then you can use a pseudo-random number generator with a fixed seed. That way you can get shuffled answers and so on, but exactly the same order each time.